File: python/nexus_connector_sdk/webhook.py
"""Webhook normalizer (SPEC-022 behavior 2) - Python binding.

The ``WebhookNormalizer`` port converts raw webhook deliveries into
canonical, versioned, correlated ``WebhookEvent`` records and verifies
signatures. Replay detection is part of the contract; a delivery that
fails signature or replay checks is rejected and never becomes an
event. Mirrors the Rust ``WebhookNormalizer`` exactly.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .vocabulary import WebhookEvent, WebhookVerification
from .wire import InvocationContext


@dataclass
class RawWebhook:
    """Raw webhook delivery before normalization."""

    raw_payload: dict[str, Any] = field(default_factory=dict)
    signature: str | None = None
    provider_event_id: str | None = None
    provider_event_type: str | None = None

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {"raw_payload": self.raw_payload}
        # Option fields serialize as JSON null (Rust serde parity).
        data["signature"] = self.signature
        data["provider_event_id"] = self.provider_event_id
        data["provider_event_type"] = self.provider_event_type
        return data

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> RawWebhook:
        return cls(
            raw_payload=dict(data.get("raw_payload", {})),
            signature=data.get("signature"),
            provider_event_id=data.get("provider_event_id"),
            provider_event_type=data.get("provider_event_type"),
        )


@dataclass
class NormalizedWebhook:
    """Normalized webhook outcome."""

    event: WebhookEvent | None = None
    verification: str = WebhookVerification.Invalid

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {"verification": self.verification}
        # Option field serializes as JSON null (Rust serde parity).
        data["event"] = self.event.to_dict() if self.event is not None else None
        return data

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> NormalizedWebhook:
        return cls(
            event=(
                WebhookEvent.from_dict(data["event"]) if data.get("event") is not None else None
            ),
            verification=str(data.get("verification", WebhookVerification.Invalid)),
        )


class WebhookNormalizer:
    """Port for normalizing and verifying webhook deliveries.

    Implementations convert a raw delivery into a canonical event,
    verify the signature and replay state, and reject deliveries that
    fail those checks. This base class carries the shared replay
    guard; subclasses supply the signature check.
    """

    def __init__(self, expected_fingerprint: str) -> None:
        self._expected_fingerprint = expected_fingerprint
        self._seen_event_ids: set[str] = set()

    def normalize(
        self,
        raw: RawWebhook,
        capability_id: str,
        context: InvocationContext,
    ) -> NormalizedWebhook:
        """Normalize a raw delivery into a canonical event, verifying
        the signature and replay state."""
        signature = raw.signature or ""
        if self._expected_fingerprint not in signature:
            return NormalizedWebhook(event=None, verification=WebhookVerification.Invalid)
        event_id = raw.provider_event_id or f"wh-{context.request_id}"
        if event_id in self._seen_event_ids:
            return NormalizedWebhook(event=None, verification=WebhookVerification.Replay)
        self._seen_event_ids.add(event_id)
        event = WebhookEvent(
            event_id=event_id,
            event_type=raw.provider_event_type or "webhook.received",
            version="1",
            correlation_id=context.correlation_id,
            payload=raw.raw_payload,
        )
        return NormalizedWebhook(event=event, verification=WebhookVerification.Valid)


class AcceptingWebhookNormalizer(WebhookNormalizer):
    """Deterministic normalizer that accepts a configured signature.

    Test/verification zone: used by the conformance corpus to prove
    the normalize contract; production webhook providers implement
    their own ``WebhookNormalizer`` with real signature verification.
    """

    pass


__all__ = [
    "AcceptingWebhookNormalizer",
    "NormalizedWebhook",
    "RawWebhook",
    "WebhookNormalizer",
]

File: crates/nexus-connector-sdk/src/webhook.rs
//! Webhook normalizer (SPEC-022 behavior 2).
//!
//! The `WebhookNormalizer` port converts raw webhook deliveries into
//! canonical, versioned, correlated `WebhookEvent` records and
//! verifies signatures. Replay detection is part of the contract;
//! a delivery that fails signature or replay checks is rejected and
//! never becomes an event.

use serde::{Deserialize, Serialize};

use nexus_capabilities::context::InvocationContext;

use crate::error::SdkError;
use crate::vocabulary::{WebhookEvent, WebhookVerification};

/// Raw webhook delivery before normalization.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RawWebhook {
    /// Provider raw payload (normalized at this boundary).
    pub raw_payload: serde_json::Value,
    /// Optional signature header content.
    pub signature: Option<String>,
    /// Optional provider event id.
    pub provider_event_id: Option<String>,
    /// Optional provider event type.
    pub provider_event_type: Option<String>,
}

/// Normalized webhook outcome.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct NormalizedWebhook {
    /// Canonical event (present only when verification passed).
    pub event: Option<WebhookEvent>,
    /// Verification result.
    pub verification: WebhookVerification,
}

/// Error produced by the webhook normalizer.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WebhookNormalizerError(pub SdkError);

impl std::fmt::Display for WebhookNormalizerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "webhook normalizer: {}", self.0)
    }
}

impl std::error::Error for WebhookNormalizerError {}

/// Port for normalizing and verifying webhook deliveries.
pub trait WebhookNormalizer: Send + Sync {
    /// Normalize a raw delivery into a canonical event, verifying the
    /// signature and replay state.
    fn normalize(
        &self,
        raw: RawWebhook,
        capability_id: String,
        context: InvocationContext,
    ) -> Result<NormalizedWebhook, WebhookNormalizerError>;
}

/// Deterministic normalizer that accepts a configured signature.
///
/// Test/verification zone: used by unit tests and the M5 conformance
/// corpus to prove the normalize contract; production webhook
/// providers implement their own `WebhookNormalizer` with real
/// signature verification.
#[derive(Debug, Clone)]
pub struct AcceptingWebhookNormalizer {
    /// Fingerprint the normalizer accepts.
    pub expected_fingerprint: String,
}

impl WebhookNormalizer for AcceptingWebhookNormalizer {
    fn normalize(
        &self,
        raw: RawWebhook,
        _capability_id: String,
        context: InvocationContext,
    ) -> Result<NormalizedWebhook, WebhookNormalizerError> {
        let signature = raw.signature.unwrap_or_default();
        let verification = if signature.contains(&self.expected_fingerprint) {
            WebhookVerification::Valid
        } else {
            WebhookVerification::Invalid
        };
        let event = if verification == WebhookVerification::Valid {
            Some(WebhookEvent {
                event_id: raw
                    .provider_event_id
                    .unwrap_or_else(|| format!("wh-{}", context.request_id)),
                event_type: raw
                    .provider_event_type
                    .unwrap_or_else(|| "webhook.received".to_string()),
                version: "1".to_string(),
                correlation_id: context.correlation_id.to_string(),
                payload: raw.raw_payload,
            })
        } else {
            None
        };
        Ok(NormalizedWebhook {
            event,
            verification,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::vocabulary::WebhookSignature;
    use nexus_domain::{CorrelationId, NexusId, PrincipalType, TenantId};

    fn ctx() -> InvocationContext {
        InvocationContext::new(
            NexusId::new("018f0f6f-9c1e-7b6e-8000-000000000001").unwrap(),
            CorrelationId::new("018f0f6f-9c1e-7b6e-8000-000000000002").unwrap(),
            None,
            "test",
            "user:alice",
            PrincipalType::Human,
            TenantId::new("018f0f6f-9c1e-7b6e-8000-000000000003").unwrap(),
            Some("mcp".to_string()),
            None,
            None,
            None,
        )
        .unwrap()
    }

    #[test]
    fn ep011_unit_webhook_normalizer_valid_signature() {
        let normalizer = AcceptingWebhookNormalizer {
            expected_fingerprint: "fp-test".to_string(),
        };
        let result = normalizer
            .normalize(
                RawWebhook {
                    raw_payload: serde_json::json!({ "ok": true }),
                    signature: Some("sha256=fp-test:abc".to_string()),
                    provider_event_id: Some("prov-1".to_string()),
                    provider_event_type: Some("invoice.paid".to_string()),
                },
                "cap.webhook".to_string(),
                ctx(),
            )
            .unwrap();
        assert_eq!(result.verification, WebhookVerification::Valid);
        let event = result.event.unwrap();
        assert_eq!(event.event_id, "prov-1");
        assert_eq!(event.event_type, "invoice.paid");
    }

    #[test]
    fn ep011_unit_webhook_normalizer_rejects_bad_signature() {
        let normalizer = AcceptingWebhookNormalizer {
            expected_fingerprint: "fp-test".to_string(),
        };
        let result = normalizer
            .normalize(
                RawWebhook {
                    raw_payload: serde_json::json!({ "ok": true }),
                    signature: Some("sha256=wrong:xyz".to_string()),
                    provider_event_id: None,
                    provider_event_type: None,
                },
                "cap.webhook".to_string(),
                ctx(),
            )
            .unwrap();
        assert_eq!(result.verification, WebhookVerification::Invalid);
        assert!(result.event.is_none());
    }

    #[test]
    fn ep011_unit_webhook_signature_never_contains_secret() {
        let sig = WebhookSignature {
            scheme: "hmac-sha256".to_string(),
            key_fingerprint: "fp-abc123".to_string(),
            value_hex: "deadbeef".to_string(),
        };
        let json = serde_json::to_value(&sig).unwrap();
        assert_eq!(json["key_fingerprint"], "fp-abc123");
        assert!(json["value_hex"].as_str().unwrap().len() <= 128);
        assert!(json.as_object().unwrap().get("key").is_none());
    }
}
